### PR TITLE
Backend/feature/lib v1 j

### PR DESCRIPTION
**Pull request by Leohemmingsson**
# Desc
* ML lib in backend
* Docker works with new changes
*

### CODE REVIEW

**Review by jacomoel — APPROVED** [review]
Really good work lads, cool that we have something to run through the application cycle (everything besides some type of feedback).

When extracting a model from the research note book, follow how the IsbitClassifierModel shared functions are implemented and extend with more specific needs in the subclass.

**Review by jacomoel — COMMENTED** [review]
We should schedule a time for moving the dimensionality reduction models into the callable lib sometime in the near future! @Varlog22

**Review by Varlog22 — APPROVED** [review]
The code for the test model looks good and seems to work. It seems based on the work done in the ML-research notebook file, and generates dataframes with appropriate fields(clustering and truth should be nice to have later) for use in the frontend. Nice work!

File: Backend/src/ml_lib/factory/getter.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from ..sub_models import QaqcTestModel
+
+
+def get_instance(model_name: str):
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/factory/getter.py:4]: what does this return?
docstring?

File: Backend/src/ml_lib/shared/shared_model.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class IsbitClassifierModel:
+
+    def __init__(
+        self, source_data_path: str | None = None, df: pd.DataFrame | None = None
+    ) -> None:
+        self._df = df
+
+    def _read_data_frame(self, file_name: str) -> pd.DataFrame:
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/shared/shared_model.py:11]: docstring

File: Backend/src/ml_lib/shared/shared_model.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        raise Exception("Not implemented.")
+
+    def latter_run(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        (mby not needed, but run after feedback loop) Latter clustering run logic, overridden by child classes.
+        """
+        raise Exception("Not implemented.")
+
+    def _df_setter(self, df: pd.DataFrame) -> None:
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/shared/shared_model.py:48]: docstring?

File: Backend/src/ml_lib/shared/shared_model.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        df = df[["text", "x", "y", "truth", "cluster"]]
+
+        # TODO: validators
+        self._df = df
+
+    @property
+    def df(self) -> pd.DataFrame:
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/shared/shared_model.py:56]: docstring?

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        zipped = list(zip(no_comma_questions, coarse_labels))
+
+        os.makedirs(os.path.dirname(output_data_path), exist_ok=True)
+        with open(output_data_path, "w", encoding="utf-8", newline="") as outfile:
+            writer = csv.writer(outfile, quoting=csv.QUOTE_MINIMAL)
+            writer.writerow(["text", "coarse label"])
+            writer.writerows(zipped)
+
+    # helper to remove outer quotation marks, specific for qaqc data set
+    def strip_outer_quotationmarks(self, q):
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:44]: what does it return

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        if q.startswith('"') and q.endswith('"'):
+            q = q[1:-1]
+        return q.strip()
+
+    # helper to remove commas in the question text, specific for qaqc data set
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:49]: why comment and not docstring as in other places

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        )
+        zipped = list(zip(no_comma_questions, coarse_labels))
+
+        os.makedirs(os.path.dirname(output_data_path), exist_ok=True)
+        with open(output_data_path, "w", encoding="utf-8", newline="") as outfile:
+            writer = csv.writer(outfile, quoting=csv.QUOTE_MINIMAL)
+            writer.writerow(["text", "coarse label"])
+            writer.writerows(zipped)
+
+    # helper to remove outer quotation marks, specific for qaqc data set
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:43]: why comment and not docstring as in other places

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    def strip_outer_quotationmarks(self, q):
+        if q.startswith('"') and q.endswith('"'):
+            q = q[1:-1]
+        return q.strip()
+
+    # helper to remove commas in the question text, specific for qaqc data set
+    def replace_commas(self, line):
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:50]: what does this return

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        pattern = r'"([^"]*?)"'
+
+        def replace_commas(match):
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:53]: same as above

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            return '"' + match.group(1).replace(",", "|") + '"'
+
+        return re.sub(pattern, replace_commas, line)
+
+    # combines the input qeustion data with the calculated 2D point data
+    def first_run(self, df: pd.DataFrame):
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:59]: check other comments

File: Backend/src/v1/run_ml.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+from fastapi import APIRouter, HTTPException
+
+router = APIRouter(
+    prefix="/run_ml",
+    tags=["Run ML"],
+)
+
+
+@router.get("/", status_code=200)
+def run(model_name: str, file: str):
```
Comment (LinusHakegard) [inline Backend/src/v1/run_ml.py:16]: return type?

File: Backend/src/validators/db_validators.py [hunk 12]
```diff
@@ -6,6 +6,7 @@ def validate_endpoint_args(func):
     For this validator to work as expected, the function arguments needs to be given as kwargs:
         e.g. func(collection=collection)
     """
+
     def wrapper_func(*args, **kwargs):
```
Comment (LinusHakegard) [inline Backend/src/validators/db_validators.py:10]: return type?

File: Backend/src/types/node.py [hunk 13]
```diff
@@ -6,12 +6,30 @@ class Node(BaseModel):
...
+                    "text": "This is a text",
+                    "x": 0.1,
+                    "y": 0.2,
+                    "truth": "true",
+                },
+            ]
+        }
+    }
+
+    def __init__(self, *args, **kwargs):
```
Comment (LinusHakegard) [inline Backend/src/types/node.py:31]: return type?

File: Backend/src/ml_lib/sub_models/qaqc_test_model.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+from sentence_transformers import SentenceTransformer
+from sklearn.cluster import KMeans
+from sklearn.decomposition import PCA
+
+from ..shared import IsbitClassifierModel
+
+
+class QaqcTestModel(IsbitClassifierModel):
+
+    def _format_data(self, file_name: str):
```
Comment (LinusHakegard) [inline Backend/src/ml_lib/sub_models/qaqc_test_model.py:15]: return type?

**Review by LinusHakegard — CHANGES_REQUESTED** [review]
Add docstrings
Add return types
A few inconsistencies

Functionality looks good.

**Review by LinusHakegard — APPROVED** [review]

**Pull request merged**